### Layered LightGBM Ensemble: Sampling from Clusters, Time-Sliced Training and Ridge Stacking

In this notebook, I build on the hierarchical clustering performed on the masked features in the feature_engineering notebook. Instead of aggregating features within each cluster, I sample directly from the clusters. Specifically, in each round I select one feature from each cluster without replacement, ensuring that no two features from the same cluster appear in a given subset. I continue this procedure until all features are sampled. This procedure is repeated until all features have been sampled.

I then apply the same time-slicing procedure, splitting the training data into 5 subsamples. On each subsample, I train multiple boosted trees using the sampled feature subsets, resulting in total of 60 boosted trees. To guard against overfitting in this noisy setting, I constrain the search space for model hyperparameters (e.g., lower ranges for bagging and feature fractions, and fewer boosted trees), encouraging simpler models that generalize better. Finally, to ensemble their outputs, I train a ridge regression model on the predictions from all trees.

In [1]:
# libraries and settings
import gc
import pandas as pd
import numpy as np
from collections import defaultdict
from scipy.stats import pearsonr
from sklearn.linear_model import Ridge
import lightgbm as lgb
import optuna
import warnings
warnings.filterwarnings('ignore')

In [2]:
df_train = pd.read_parquet('../data/input/train.parquet')
df_test = pd.read_parquet('../data/input/test.parquet')

#### Sampling Features
After the feature sampling procedure is complete, feature samples that have less than 10 features are dropped. Moreover, the orderbook features is added to all masked feature subsets to be used in the training procedure. In the end, Iend up with 12 feature subsets.

In [3]:
# Clustering the features
X_cols = df_train.columns[df_train.columns.str.startswith('X')]
clusters = pd.Series(pd.read_csv('../data/intermediate/clusters.csv', index_col = 0).iloc[:,0])

# shuffle the features
clusters = clusters.sample(frac = 1, random_state = 42)

# sampling teh clusters
clusters_dict = defaultdict(list)

for c in np.unique(clusters):
    X = clusters[clusters == c]
    clusters_dict[c] = X.index.tolist()

total = 0
features_list = []
n_features = len(clusters)

while total < n_features:
    features = []
    for c in clusters_dict.keys():
        if len(clusters_dict[c]) > 0:
            features.append(clusters_dict[c][-1])
            clusters_dict[c].pop()
            total += 1
    features_list.append(features)
    features = list([])
    
while len(features_list[-1]) < 10:
    features_list.pop()

for i in range(len(features_list)):
    features_list[i] = ['bid_qty', 'ask_qty', 'buy_qty', 'sell_qty', 'volume'] + features_list[i]

#### Ensemble Prediction Model
I design a class to pipeline the tuning and training procedures. First, the training and validation datasets are split into 5 subsamples by selecting every 5th row (after sorting teh data based on their timestamps) into each subsample. For each subsample, I train boosted trees on the corresponding training data and feature subsets, tuning them on their paired validation subsample using the correlation metric. The correlation metric is chosen because it was the evaluation metric used in the competition.

Note that because the data is highly noisy, I restrict the search space for the bagging fraction and feature fraction to the lower ranges, [0.2, 0.6] and [0.1,0.5], respectively. Likewise, the number of boosted trees is tuned within the range of 200 to 500, which is also on the lower end. These choices are intended to promote better generalization in the presence of noisy inputs.

Once all 60 tuned trees are trained, I generate predictions from them on the full training and validation datasets. These stacked predictions form the inputs to a ridge regression model, which is trained on the training predictions. The ridge regularization parameter is tuned based on the correlation metric evaluated on the validation predictions.

Finally, after identifying the optimal ridge parameter, I retrain all boosted trees with their tuned hyperparameters on the full dataset (including the validation sample). I then refit the ridge regression using the optimized regularization parameter on the stacked predictions from the full data. This final ensemble model is used to produce predictions on the test dataset.

In [5]:
""" **************** ENSEMBLE MODEL ****************** """
class LightGBMTimeSeriesEnsemble:
    def __init__(self, n_slices, n_trials, feature_sets, alphas=None,
                 metric="l2", random_seed=42):
        """
        n_slices  : number of time slices (e.g. in our case 20)
        n_trials  : Optuna trials per learner
        feature_sets: sets of features that are each separately used for training trees.
        alphas    : candidate ridge regularization strengths
        metric    : LightGBM eval metric (for early stopping)
        """
        self.n_slices = n_slices
        self.n_trials = n_trials
        self.feature_sets = feature_sets
        self.metric = metric
        self.alphas = alphas if alphas is not None else np.logspace(-2, 5, 30)
        self.seed = random_seed

        self.best_params_list = []
        self.best_iterations = []
        self.models = []
        self.ridge = None
        self.best_alpha = None

    def _optuna_objective(self, trial, X_tr, y_tr, X_val, y_val):
        params = {
            'objective': 'regression',
            'metric': 'rmse',  
            'verbosity': -1,
            'learning_rate': trial.suggest_float('learning_rate', 0.005, 0.1, log=True),
            'num_leaves': trial.suggest_int('num_leaves', 16, 64),
            'max_depth': trial.suggest_int('max_depth', 2, 6),
            'min_data_in_leaf': trial.suggest_int('min_data_in_leaf', 10, 100),
            'feature_fraction': trial.suggest_float('feature_fraction', 0.1, 0.5),
            'bagging_fraction': trial.suggest_float('bagging_fraction', 0.2, 0.6),
            'bagging_freq': trial.suggest_int('bagging_freq', 1, 10),
            'lambda_l1': trial.suggest_float('lambda_l1', 0.0, 5.0),
            'lambda_l2': trial.suggest_float('lambda_l2', 0.0, 5.0),
            'num_boost_round': trial.suggest_int('num_boost_round', 200,500),
            'early_stopping_rounds': 50
        }
        model = lgb.train(
            params,
            lgb.Dataset(X_tr, y_tr),
            valid_sets=[lgb.Dataset(X_val, y_val)]
        )
        pred = model.predict(X_val)
        corr, _ = pearsonr(pred, y_val)
        return -corr

    def fit(self, X_train, y_train, X_val, y_val):
        """
        1) Fit each base model on X_train[i::n] versus y_train[i::n] with Optuna tuning.
        Next steps are carried out in the fit_ridge function:
        2) Make predictions of each base model on *full* X_train => (n_samples_train x n_slices)
        3) Ridge-fit on (train_preds, y_train)
        4) Select alpha by evaluating ridge on full X_val predictions
        """
        self.models = []
        self.best_params_list = []
        self.best_iterations = []

        for j in range(len(self.feature_sets)):
            for i in range(self.n_slices):
                print(f'Training on feature set {j+1}/{len(self.feature_sets)}, model {i+1}/{self.n_slices}')
                
                Xt = X_train.iloc[i::self.n_slices][self.feature_sets[j]]
                yt = y_train.iloc[i::self.n_slices]
                Xv = X_val.iloc[i::self.n_slices][self.feature_sets[j]]
                yv = y_val.iloc[i::self.n_slices]
    
                # Tune
                sampler = optuna.samplers.TPESampler(seed=self.seed)
                study = optuna.create_study(direction='minimize', sampler=sampler)
                study.optimize(
                    lambda trial: self._optuna_objective(trial, Xt, yt, Xv, yv),
                    n_trials=self.n_trials,
                )
                best_params = study.best_params
                best_params['seed'] = self.seed
                
    
                # Train model with best params
                model = lgb.train(
                    {
                    **best_params, 
                    'objective':'regression',
                    'metric':self.metric,
                    'verbosity':-1, 
                    'early_stopping_rounds': 50
                    },
                    lgb.Dataset(Xt, yt),
                    valid_sets=[lgb.Dataset(Xv, yv)]
                )
                best_params['num_boost_round'] = model.best_iteration
                self.best_params_list.append(best_params)
                self.best_iterations.append(model.best_iteration)
                self.models.append(model)

        self.fit_ridge(X_train, y_train, X_val, y_val, self.alphas)
        
        
    def fit_ridge(self, X_train, y_train, X_val, y_val, alphas):
        """
        1) Make predictions of each base model on *full* X_train => (n_samples_train x n_slices)
        2) Ridge-fit on (train_preds, y_train)
        3) Select alpha by evaluating ridge on full X_val predictions
        """
        train_preds = []
        val_preds   = []
        idx = 0

         # --- predictions on full training and validation set ---
        for subset in self.feature_sets:
            for _ in range(self.n_slices):
                m  = self.models[idx]
                it = self.best_iterations[idx]
                train_preds.append(m.predict(X_train[subset], num_iteration=it))
                val_preds.append(m.predict(X_val[subset],   num_iteration=it))
                idx += 1
        
        train_stack = np.column_stack(train_preds)
        val_stack   = np.column_stack(val_preds)
       
        self.alphas = alphas
        best_corr = -np.inf
        for a in self.alphas:
            r = Ridge(alpha=a)
            r.fit(train_stack, y_train)
            corr, _ = pearsonr(r.predict(val_stack), y_val)
            print(f' ridge alpha: {a:.5f}, val corr: {corr:.5f}')
            if corr > best_corr:
                best_corr, self.best_alpha, self.ridge = corr, a, r

        print(f'Best ridge alpha: {self.best_alpha:.5f}, val corr: {best_corr:.5f}')
    
    
    def refit_full(self, X_full, y_full):
        """
        Retrain base models and the ridge regression model on full data using best params and alpha.
        """
        self.models = []
        idx = 0
        for subset in self.feature_sets:
            for _ in range(self.n_slices):
                print(f'Refitting model {idx+1}/{self.n_slices*len(self.feature_sets)} on full data...')
                params  = self.best_params_list[idx]
                model = lgb.train(
                    {
                    **params,  
                    'objective':'regression',
                    'metric':self.metric,
                    'verbosity':-1
                    },
                    lgb.Dataset(X_full[subset], y_full)
                )
                self.models.append(model)
                idx += 1
        
        preds = []
        idx = 0
         # --- predictions on full dataset
        for subset in self.feature_sets:
            for _ in range(self.n_slices):
                m  = self.models[idx]
                it = self.best_iterations[idx]
                preds.append(m.predict(X_full[subset], num_iteration=it))
                idx += 1
                
        full_stack = stack = np.column_stack(preds)       
        self.ridge = Ridge(alpha=self.best_alpha)
        self.ridge.fit(full_stack, y_full)

    def predict(self, X):
        """
        Predict stacked output on full dataset X
        """
        preds = []
        idx = 0
        for subset in self.feature_sets:
            for _ in range(self.n_slices):
                m  = self.models[idx]
                it = self.best_iterations[idx]
                preds.append(m.predict(X[subset], num_iteration=it))
                idx += 1
        stack = np.column_stack(preds)
        
        return self.ridge.predict(stack)



Due to the time-series nature of the data, I sort observations by their timestamps to prevent any look-ahead bias. The first 80% of the data is used for training, while the remaining 20% is reserved for validation.

In [6]:
df_train.sort_index(inplace = True)
n_train = round(df_train.shape[0]*0.8)
X_train = df_train.iloc[:n_train].drop(columns = ['label'])
X_val = df_train.iloc[n_train:].drop(columns = ['label'])
y_train = df_train.iloc[:n_train]['label'] 
y_val = df_train.iloc[n_train:]['label'] 

ensemble = LightGBMTimeSeriesEnsemble(n_slices=5, n_trials=50, feature_sets=features_list)
ensemble.fit(X_train, y_train, X_val, y_val)

[I 2025-08-19 14:47:32,047] A new study created in memory with name: no-name-94a59bad-11a4-4401-87c3-51df3e531bc1


Training on feature set 1/12, model 1/5


[I 2025-08-19 14:47:33,925] Trial 0 finished with value: -0.08080226745370152 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.08080226745370152.
[I 2025-08-19 14:47:35,966] Trial 1 finished with value: -0.0829087427784682 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.0829087427784682.
[I 2025-08-19 14:47:37,512] Trial 2 finished with value: -0.08354512870665791 and parameters: {'learning_rate': 0.03126143958203108, 'num

Training on feature set 1/12, model 2/5


[I 2025-08-19 14:49:10,237] A new study created in memory with name: no-name-2a66fce1-b68f-4873-af17-66419fd7296d
[I 2025-08-19 14:49:11,710] Trial 0 finished with value: -0.07518809983599434 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.07518809983599434.
[I 2025-08-19 14:49:14,105] Trial 1 finished with value: -0.09776442774770872 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.09776442774770872.
[I 2025-08-19 14:49:1

Training on feature set 1/12, model 3/5


[I 2025-08-19 14:51:01,264] A new study created in memory with name: no-name-e1dab572-8944-4fcc-bc7e-d4366ed66f14
[I 2025-08-19 14:51:03,183] Trial 0 finished with value: -0.07490628080181352 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.07490628080181352.
[I 2025-08-19 14:51:05,442] Trial 1 finished with value: -0.11277886291191724 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.11277886291191724.
[I 2025-08-19 14:51:0

Training on feature set 1/12, model 4/5


[I 2025-08-19 14:52:30,188] Trial 0 finished with value: -0.07308314029350702 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.07308314029350702.
[I 2025-08-19 14:52:31,906] Trial 1 finished with value: -0.10564251110475886 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.10564251110475886.
[I 2025-08-19 14:52:33,254] Trial 2 finished with value: -0.08097530750941989 and parameters: {'learning_rate': 0.03126143958203108, 'n

Training on feature set 1/12, model 5/5


[I 2025-08-19 14:53:52,496] Trial 0 finished with value: -0.06665103380358987 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.06665103380358987.
[I 2025-08-19 14:53:54,137] Trial 1 finished with value: -0.08393692109649556 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.08393692109649556.
[I 2025-08-19 14:53:55,366] Trial 2 finished with value: -0.08439412016811694 and parameters: {'learning_rate': 0.03126143958203108, 'n

Training on feature set 2/12, model 1/5


[I 2025-08-19 14:55:03,389] Trial 0 finished with value: -0.03683947911334515 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.03683947911334515.
[I 2025-08-19 14:55:04,405] Trial 1 finished with value: -0.04737348007491928 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.04737348007491928.
[I 2025-08-19 14:55:05,297] Trial 2 finished with value: -0.04257209270202353 and parameters: {'learning_rate': 0.03126143958203108, 'n

Training on feature set 2/12, model 2/5


[I 2025-08-19 14:56:00,216] Trial 0 finished with value: -0.0427818568630155 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.0427818568630155.
[I 2025-08-19 14:56:01,616] Trial 1 finished with value: -0.04138803929985441 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 0 with value: -0.0427818568630155.
[I 2025-08-19 14:56:02,501] Trial 2 finished with value: -0.04863318623790251 and parameters: {'learning_rate': 0.03126143958203108, 'num_

Training on feature set 2/12, model 3/5


[I 2025-08-19 14:56:51,585] Trial 0 finished with value: -0.058383304031284444 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.058383304031284444.
[I 2025-08-19 14:56:52,830] Trial 1 finished with value: -0.048057266488979025 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 0 with value: -0.058383304031284444.
[I 2025-08-19 14:56:53,708] Trial 2 finished with value: -0.033453833759745776 and parameters: {'learning_rate': 0.0312614395820310

Training on feature set 2/12, model 4/5


[I 2025-08-19 14:57:50,441] Trial 0 finished with value: -0.033614677809269314 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.033614677809269314.
[I 2025-08-19 14:57:51,548] Trial 1 finished with value: -0.04615085889792399 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.04615085889792399.
[I 2025-08-19 14:57:52,686] Trial 2 finished with value: -0.04719519619602489 and parameters: {'learning_rate': 0.03126143958203108, 

Training on feature set 2/12, model 5/5


[I 2025-08-19 14:58:51,897] Trial 0 finished with value: -0.024168096612199604 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.024168096612199604.
[I 2025-08-19 14:58:53,155] Trial 1 finished with value: -0.037934214869067125 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.037934214869067125.
[I 2025-08-19 14:58:54,152] Trial 2 finished with value: -0.04815460330344358 and parameters: {'learning_rate': 0.03126143958203108

Training on feature set 3/12, model 1/5


[I 2025-08-19 14:59:47,108] Trial 0 finished with value: -0.04445493508201475 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.04445493508201475.
[I 2025-08-19 14:59:48,009] Trial 1 finished with value: -0.06939055786423222 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.06939055786423222.
[I 2025-08-19 14:59:48,777] Trial 2 finished with value: -0.053960029145192316 and parameters: {'learning_rate': 0.03126143958203108, '

Training on feature set 3/12, model 2/5


[I 2025-08-19 15:00:55,977] Trial 0 finished with value: -0.06218539508214488 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.06218539508214488.
[I 2025-08-19 15:00:57,324] Trial 1 finished with value: -0.06512459586665434 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.06512459586665434.
[I 2025-08-19 15:00:58,567] Trial 2 finished with value: -0.0783350357366696 and parameters: {'learning_rate': 0.03126143958203108, 'nu

Training on feature set 3/12, model 3/5


[I 2025-08-19 15:01:58,676] Trial 0 finished with value: -0.04628771769981871 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.04628771769981871.
[I 2025-08-19 15:01:59,846] Trial 1 finished with value: -0.0410100481294726 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 0 with value: -0.04628771769981871.
[I 2025-08-19 15:02:00,765] Trial 2 finished with value: -0.07497533562362048 and parameters: {'learning_rate': 0.03126143958203108, 'nu

Training on feature set 3/12, model 4/5


[I 2025-08-19 15:02:59,634] Trial 0 finished with value: -0.06585603929916316 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.06585603929916316.
[I 2025-08-19 15:03:00,927] Trial 1 finished with value: -0.050412249538237704 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 0 with value: -0.06585603929916316.
[I 2025-08-19 15:03:01,628] Trial 2 finished with value: -0.06430639022414225 and parameters: {'learning_rate': 0.03126143958203108, '

Training on feature set 3/12, model 5/5


[I 2025-08-19 15:03:47,180] Trial 0 finished with value: -0.048603788582499835 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.048603788582499835.
[I 2025-08-19 15:03:48,459] Trial 1 finished with value: -0.03870736321725069 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 0 with value: -0.048603788582499835.
[I 2025-08-19 15:03:49,382] Trial 2 finished with value: -0.07560974567131634 and parameters: {'learning_rate': 0.03126143958203108,

Training on feature set 4/12, model 1/5


[I 2025-08-19 15:04:44,317] Trial 0 finished with value: -0.06623789744220616 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.06623789744220616.
[I 2025-08-19 15:04:45,531] Trial 1 finished with value: -0.049379455570654755 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 0 with value: -0.06623789744220616.
[I 2025-08-19 15:04:46,320] Trial 2 finished with value: -0.08049529195950031 and parameters: {'learning_rate': 0.03126143958203108, '

Training on feature set 4/12, model 2/5


[I 2025-08-19 15:05:34,300] Trial 0 finished with value: -0.07846894656338074 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.07846894656338074.
[I 2025-08-19 15:05:35,698] Trial 1 finished with value: -0.06612603292430803 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 0 with value: -0.07846894656338074.
[I 2025-08-19 15:05:36,466] Trial 2 finished with value: -0.06948700807858542 and parameters: {'learning_rate': 0.03126143958203108, 'n

Training on feature set 4/12, model 3/5


[I 2025-08-19 15:06:18,946] Trial 0 finished with value: -0.058780701447198866 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.058780701447198866.
[I 2025-08-19 15:06:20,702] Trial 1 finished with value: -0.05973186109375739 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.05973186109375739.
[I 2025-08-19 15:06:21,449] Trial 2 finished with value: -0.08621795676826914 and parameters: {'learning_rate': 0.03126143958203108, 

Training on feature set 4/12, model 4/5


[I 2025-08-19 15:07:04,275] Trial 0 finished with value: -0.05239148785393869 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.05239148785393869.
[I 2025-08-19 15:07:05,470] Trial 1 finished with value: -0.052370754813034756 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 0 with value: -0.05239148785393869.
[I 2025-08-19 15:07:06,197] Trial 2 finished with value: -0.07480600428465325 and parameters: {'learning_rate': 0.03126143958203108, '

Training on feature set 4/12, model 5/5


[I 2025-08-19 15:07:46,914] Trial 0 finished with value: -0.050571226194586344 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.050571226194586344.
[I 2025-08-19 15:07:48,289] Trial 1 finished with value: -0.052158644184769495 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.052158644184769495.
[I 2025-08-19 15:07:49,097] Trial 2 finished with value: -0.07495212175354246 and parameters: {'learning_rate': 0.03126143958203108

Training on feature set 5/12, model 1/5


[I 2025-08-19 15:08:31,100] Trial 0 finished with value: -0.027782276003165747 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.027782276003165747.
[I 2025-08-19 15:08:32,054] Trial 1 finished with value: -0.03546451070973794 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.03546451070973794.
[I 2025-08-19 15:08:32,705] Trial 2 finished with value: -0.09785217538399987 and parameters: {'learning_rate': 0.03126143958203108, 

Training on feature set 5/12, model 2/5


[I 2025-08-19 15:09:05,040] Trial 0 finished with value: -0.023326982006518907 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.023326982006518907.
[I 2025-08-19 15:09:05,894] Trial 1 finished with value: -0.030518683663448613 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.030518683663448613.
[I 2025-08-19 15:09:06,378] Trial 2 finished with value: -0.06822381261702518 and parameters: {'learning_rate': 0.03126143958203108

Training on feature set 5/12, model 3/5


[I 2025-08-19 15:09:43,926] Trial 0 finished with value: -0.02843593123072474 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.02843593123072474.
[I 2025-08-19 15:09:44,763] Trial 1 finished with value: -0.03879168712506021 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.03879168712506021.
[I 2025-08-19 15:09:45,229] Trial 2 finished with value: -0.0870950321614608 and parameters: {'learning_rate': 0.03126143958203108, 'nu

Training on feature set 5/12, model 4/5


[I 2025-08-19 15:10:14,987] Trial 0 finished with value: -0.020624534038919506 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.020624534038919506.
[I 2025-08-19 15:10:15,803] Trial 1 finished with value: -0.025536826524847955 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.025536826524847955.
[I 2025-08-19 15:10:16,256] Trial 2 finished with value: -0.08322046949135556 and parameters: {'learning_rate': 0.03126143958203108

Training on feature set 5/12, model 5/5


[I 2025-08-19 15:10:43,856] Trial 0 finished with value: -0.01204638352789024 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.01204638352789024.
[I 2025-08-19 15:10:44,703] Trial 1 finished with value: -0.010968956131080218 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 0 with value: -0.01204638352789024.
[I 2025-08-19 15:10:45,268] Trial 2 finished with value: -0.060232667587175395 and parameters: {'learning_rate': 0.03126143958203108, 

Training on feature set 6/12, model 1/5


[I 2025-08-19 15:11:19,559] Trial 0 finished with value: -0.04708736275447894 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.04708736275447894.
[I 2025-08-19 15:11:20,389] Trial 1 finished with value: -0.07411378448637945 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.07411378448637945.
[I 2025-08-19 15:11:20,894] Trial 2 finished with value: -0.03722769571969739 and parameters: {'learning_rate': 0.03126143958203108, 'n

Training on feature set 6/12, model 2/5


[I 2025-08-19 15:11:55,338] Trial 0 finished with value: -0.055018991303925374 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.055018991303925374.
[I 2025-08-19 15:11:55,983] Trial 1 finished with value: -0.04749820764615846 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 0 with value: -0.055018991303925374.
[I 2025-08-19 15:11:56,555] Trial 2 finished with value: -0.03683926495073053 and parameters: {'learning_rate': 0.03126143958203108,

Training on feature set 6/12, model 3/5


[I 2025-08-19 15:12:28,306] Trial 0 finished with value: -0.03969071596242721 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.03969071596242721.
[I 2025-08-19 15:12:29,138] Trial 1 finished with value: -0.061119160309545284 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.061119160309545284.
[I 2025-08-19 15:12:29,617] Trial 2 finished with value: -0.04959829241210686 and parameters: {'learning_rate': 0.03126143958203108, 

Training on feature set 6/12, model 4/5


[I 2025-08-19 15:12:52,334] Trial 0 finished with value: -0.026732667459268285 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.026732667459268285.
[I 2025-08-19 15:12:53,112] Trial 1 finished with value: -0.052876133298900826 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.052876133298900826.
[I 2025-08-19 15:12:53,521] Trial 2 finished with value: -0.056425625990056794 and parameters: {'learning_rate': 0.0312614395820310

Training on feature set 6/12, model 5/5


[I 2025-08-19 15:13:16,475] Trial 0 finished with value: -0.03819020468857306 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.03819020468857306.
[I 2025-08-19 15:13:17,048] Trial 1 finished with value: -0.06556975643345545 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.06556975643345545.
[I 2025-08-19 15:13:17,522] Trial 2 finished with value: -0.044232022097461625 and parameters: {'learning_rate': 0.03126143958203108, '

Training on feature set 7/12, model 1/5


[I 2025-08-19 15:13:46,673] Trial 0 finished with value: -0.039155353438052896 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.039155353438052896.
[I 2025-08-19 15:13:47,576] Trial 1 finished with value: -0.05947813117095212 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.05947813117095212.
[I 2025-08-19 15:13:47,881] Trial 2 finished with value: -0.049024433401390476 and parameters: {'learning_rate': 0.03126143958203108,

Training on feature set 7/12, model 2/5


[I 2025-08-19 15:14:10,136] Trial 0 finished with value: -0.03890860617946148 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.03890860617946148.
[I 2025-08-19 15:14:10,791] Trial 1 finished with value: -0.062273998641924774 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.062273998641924774.
[I 2025-08-19 15:14:11,101] Trial 2 finished with value: -0.05287721154274722 and parameters: {'learning_rate': 0.03126143958203108, 

Training on feature set 7/12, model 3/5


[I 2025-08-19 15:14:34,324] Trial 0 finished with value: -0.05261409019975308 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.05261409019975308.
[I 2025-08-19 15:14:34,973] Trial 1 finished with value: -0.05280838986218275 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.05280838986218275.
[I 2025-08-19 15:14:35,273] Trial 2 finished with value: -0.056341886707923575 and parameters: {'learning_rate': 0.03126143958203108, '

Training on feature set 7/12, model 4/5


[I 2025-08-19 15:14:54,862] Trial 0 finished with value: -0.04309034425919185 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.04309034425919185.
[I 2025-08-19 15:14:55,685] Trial 1 finished with value: -0.0630686494914422 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.0630686494914422.
[I 2025-08-19 15:14:56,071] Trial 2 finished with value: -0.05544886981114232 and parameters: {'learning_rate': 0.03126143958203108, 'num

Training on feature set 7/12, model 5/5


[I 2025-08-19 15:15:17,767] Trial 0 finished with value: -0.03479948819284928 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.03479948819284928.
[I 2025-08-19 15:15:18,599] Trial 1 finished with value: -0.054517763080092874 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.054517763080092874.
[I 2025-08-19 15:15:18,897] Trial 2 finished with value: -0.04112559654975899 and parameters: {'learning_rate': 0.03126143958203108, 

Training on feature set 8/12, model 1/5


[I 2025-08-19 15:15:38,186] Trial 0 finished with value: -0.06786033767120789 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.06786033767120789.
[I 2025-08-19 15:15:38,537] Trial 1 finished with value: -0.05310884051185621 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 0 with value: -0.06786033767120789.
[I 2025-08-19 15:15:38,815] Trial 2 finished with value: -0.05884798981190565 and parameters: {'learning_rate': 0.03126143958203108, 'n

Training on feature set 8/12, model 2/5


[I 2025-08-19 15:15:55,871] Trial 0 finished with value: -0.05453366034943423 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.05453366034943423.
[I 2025-08-19 15:15:56,279] Trial 1 finished with value: -0.05483555460103591 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.05483555460103591.
[I 2025-08-19 15:15:56,595] Trial 2 finished with value: -0.05648966339404464 and parameters: {'learning_rate': 0.03126143958203108, 'n

Training on feature set 8/12, model 3/5


[I 2025-08-19 15:16:12,780] Trial 0 finished with value: -0.06315053183588806 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.06315053183588806.
[I 2025-08-19 15:16:13,201] Trial 1 finished with value: -0.056716315320668446 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 0 with value: -0.06315053183588806.
[I 2025-08-19 15:16:13,524] Trial 2 finished with value: -0.06635822546156411 and parameters: {'learning_rate': 0.03126143958203108, '

Training on feature set 8/12, model 4/5


[I 2025-08-19 15:16:29,212] Trial 0 finished with value: -0.06784649328478742 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.06784649328478742.
[I 2025-08-19 15:16:29,599] Trial 1 finished with value: -0.053874571577962244 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 0 with value: -0.06784649328478742.
[I 2025-08-19 15:16:29,962] Trial 2 finished with value: -0.07233891089826619 and parameters: {'learning_rate': 0.03126143958203108, '

Training on feature set 8/12, model 5/5


[I 2025-08-19 15:16:46,389] Trial 0 finished with value: -0.06654569408741232 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.06654569408741232.
[I 2025-08-19 15:16:46,741] Trial 1 finished with value: -0.05517441683352222 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 0 with value: -0.06654569408741232.
[I 2025-08-19 15:16:46,973] Trial 2 finished with value: -0.044897941081644376 and parameters: {'learning_rate': 0.03126143958203108, '

Training on feature set 9/12, model 1/5


[I 2025-08-19 15:17:02,899] Trial 0 finished with value: -0.04143942866223918 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.04143942866223918.
[I 2025-08-19 15:17:03,551] Trial 1 finished with value: -0.05496498310277467 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.05496498310277467.
[I 2025-08-19 15:17:03,798] Trial 2 finished with value: -0.05967697459607364 and parameters: {'learning_rate': 0.03126143958203108, 'n

Training on feature set 9/12, model 2/5


[I 2025-08-19 15:17:20,642] Trial 0 finished with value: -0.0558013461018931 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.0558013461018931.
[I 2025-08-19 15:17:20,981] Trial 1 finished with value: -0.07446314546777652 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.07446314546777652.
[I 2025-08-19 15:17:21,167] Trial 2 finished with value: -0.034521327595688475 and parameters: {'learning_rate': 0.03126143958203108, 'nu

Training on feature set 9/12, model 3/5


[I 2025-08-19 15:17:37,905] Trial 0 finished with value: -0.046831682030173474 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.046831682030173474.
[I 2025-08-19 15:17:38,262] Trial 1 finished with value: -0.04255632748757936 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 0 with value: -0.046831682030173474.
[I 2025-08-19 15:17:38,457] Trial 2 finished with value: -0.04904877958282526 and parameters: {'learning_rate': 0.03126143958203108,

Training on feature set 9/12, model 4/5


[I 2025-08-19 15:17:53,992] Trial 0 finished with value: -0.03655211885216314 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.03655211885216314.
[I 2025-08-19 15:17:54,452] Trial 1 finished with value: -0.042591735375466726 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.042591735375466726.
[I 2025-08-19 15:17:54,664] Trial 2 finished with value: -0.03722675986197502 and parameters: {'learning_rate': 0.03126143958203108, 

Training on feature set 9/12, model 5/5


[I 2025-08-19 15:18:11,785] Trial 0 finished with value: -0.05133261822362868 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.05133261822362868.
[I 2025-08-19 15:18:12,099] Trial 1 finished with value: -0.049926246164457205 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 0 with value: -0.05133261822362868.
[I 2025-08-19 15:18:12,353] Trial 2 finished with value: -0.020579421925486382 and parameters: {'learning_rate': 0.03126143958203108, 

Training on feature set 10/12, model 1/5


[I 2025-08-19 15:18:26,380] Trial 0 finished with value: -0.051522722424680874 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.051522722424680874.
[I 2025-08-19 15:18:26,795] Trial 1 finished with value: -0.0541280591990264 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.0541280591990264.
[I 2025-08-19 15:18:26,960] Trial 2 finished with value: -0.07384808150339664 and parameters: {'learning_rate': 0.03126143958203108, 'n

Training on feature set 10/12, model 2/5


[I 2025-08-19 15:18:38,534] Trial 0 finished with value: -0.0492876087289323 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.0492876087289323.
[I 2025-08-19 15:18:39,141] Trial 1 finished with value: -0.051646159230779955 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.051646159230779955.
[I 2025-08-19 15:18:39,331] Trial 2 finished with value: -0.06622917387696384 and parameters: {'learning_rate': 0.03126143958203108, 'n

Training on feature set 10/12, model 3/5


[I 2025-08-19 15:18:50,642] Trial 0 finished with value: -0.05332780032600776 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.05332780032600776.
[I 2025-08-19 15:18:51,069] Trial 1 finished with value: -0.0472330351933899 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 0 with value: -0.05332780032600776.
[I 2025-08-19 15:18:51,244] Trial 2 finished with value: -0.07367842404103422 and parameters: {'learning_rate': 0.03126143958203108, 'nu

Training on feature set 10/12, model 4/5


[I 2025-08-19 15:19:05,300] Trial 0 finished with value: -0.056902273123858114 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.056902273123858114.
[I 2025-08-19 15:19:05,834] Trial 1 finished with value: -0.05050317626672374 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 0 with value: -0.056902273123858114.
[I 2025-08-19 15:19:06,018] Trial 2 finished with value: -0.0768624564656249 and parameters: {'learning_rate': 0.03126143958203108, 

Training on feature set 10/12, model 5/5


[I 2025-08-19 15:19:19,980] Trial 0 finished with value: -0.046096427470252166 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.046096427470252166.
[I 2025-08-19 15:19:20,554] Trial 1 finished with value: -0.042891845078698206 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 0 with value: -0.046096427470252166.
[I 2025-08-19 15:19:20,750] Trial 2 finished with value: -0.07458880875031235 and parameters: {'learning_rate': 0.03126143958203108

Training on feature set 11/12, model 1/5


[I 2025-08-19 15:19:35,148] Trial 0 finished with value: -0.04957998029258096 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.04957998029258096.
[I 2025-08-19 15:19:35,812] Trial 1 finished with value: -0.06789117645432509 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.06789117645432509.
[I 2025-08-19 15:19:36,153] Trial 2 finished with value: -0.09197778057954646 and parameters: {'learning_rate': 0.03126143958203108, 'n

Training on feature set 11/12, model 2/5


[I 2025-08-19 15:19:57,182] Trial 0 finished with value: -0.052240604425505474 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.052240604425505474.
[I 2025-08-19 15:19:58,121] Trial 1 finished with value: -0.06831493801735983 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.06831493801735983.
[I 2025-08-19 15:19:58,564] Trial 2 finished with value: -0.10032318175334759 and parameters: {'learning_rate': 0.03126143958203108, 

Training on feature set 11/12, model 3/5


[I 2025-08-19 15:20:19,847] Trial 0 finished with value: -0.05609038978352243 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.05609038978352243.
[I 2025-08-19 15:20:20,581] Trial 1 finished with value: -0.07072822228060847 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.07072822228060847.
[I 2025-08-19 15:20:20,924] Trial 2 finished with value: -0.09179614612813947 and parameters: {'learning_rate': 0.03126143958203108, 'n

Training on feature set 11/12, model 4/5


[I 2025-08-19 15:20:37,292] Trial 0 finished with value: -0.0585336659156949 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.0585336659156949.
[I 2025-08-19 15:20:37,758] Trial 1 finished with value: -0.07193518748432712 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.07193518748432712.
[I 2025-08-19 15:20:38,103] Trial 2 finished with value: -0.09541560588912265 and parameters: {'learning_rate': 0.03126143958203108, 'num

Training on feature set 11/12, model 5/5


[I 2025-08-19 15:20:57,614] Trial 0 finished with value: -0.0503245615584003 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.0503245615584003.
[I 2025-08-19 15:20:58,257] Trial 1 finished with value: -0.06388551850675556 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.06388551850675556.
[I 2025-08-19 15:20:58,549] Trial 2 finished with value: -0.08015204706709556 and parameters: {'learning_rate': 0.03126143958203108, 'num

Training on feature set 12/12, model 1/5


[I 2025-08-19 15:21:15,825] Trial 0 finished with value: -0.03547719002577322 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.03547719002577322.
[I 2025-08-19 15:21:16,573] Trial 1 finished with value: -0.040094639321861754 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.040094639321861754.
[I 2025-08-19 15:21:16,790] Trial 2 finished with value: -0.03448935912543291 and parameters: {'learning_rate': 0.03126143958203108, 

Training on feature set 12/12, model 2/5


[I 2025-08-19 15:21:33,067] Trial 0 finished with value: -0.037700909320489154 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.037700909320489154.
[I 2025-08-19 15:21:33,663] Trial 1 finished with value: -0.03932883564655993 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.03932883564655993.
[I 2025-08-19 15:21:33,835] Trial 2 finished with value: -0.041407201081876205 and parameters: {'learning_rate': 0.03126143958203108,

Training on feature set 12/12, model 3/5


[I 2025-08-19 15:21:48,770] Trial 0 finished with value: -0.03069009167204721 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.03069009167204721.
[I 2025-08-19 15:21:49,365] Trial 1 finished with value: -0.03775534419998295 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.03775534419998295.
[I 2025-08-19 15:21:49,578] Trial 2 finished with value: -0.04418261400450694 and parameters: {'learning_rate': 0.03126143958203108, 'n

Training on feature set 12/12, model 4/5


[I 2025-08-19 15:22:03,496] Trial 0 finished with value: -0.03140718341285065 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.03140718341285065.
[I 2025-08-19 15:22:03,983] Trial 1 finished with value: -0.04586556494130817 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 1 with value: -0.04586556494130817.
[I 2025-08-19 15:22:04,175] Trial 2 finished with value: -0.05599326166323471 and parameters: {'learning_rate': 0.03126143958203108, 'n

Training on feature set 12/12, model 5/5


[I 2025-08-19 15:22:15,709] Trial 0 finished with value: -0.035482624352066464 and parameters: {'learning_rate': 0.015355286838886862, 'num_leaves': 62, 'max_depth': 5, 'min_data_in_leaf': 64, 'feature_fraction': 0.1624074561769746, 'bagging_fraction': 0.26239780813448105, 'bagging_freq': 1, 'lambda_l1': 4.330880728874676, 'lambda_l2': 3.005575058716044, 'num_boost_round': 413}. Best is trial 0 with value: -0.035482624352066464.
[I 2025-08-19 15:22:16,018] Trial 1 finished with value: -0.02281703646965043 and parameters: {'learning_rate': 0.005318033256270142, 'num_leaves': 63, 'max_depth': 6, 'min_data_in_leaf': 29, 'feature_fraction': 0.17272998688284025, 'bagging_fraction': 0.27336180394137355, 'bagging_freq': 4, 'lambda_l1': 2.6237821581611893, 'lambda_l2': 2.1597250932105787, 'num_boost_round': 287}. Best is trial 0 with value: -0.035482624352066464.
[I 2025-08-19 15:22:16,178] Trial 2 finished with value: -0.03693647326672452 and parameters: {'learning_rate': 0.03126143958203108,

 ridge alpha: 0.01000, val corr: 0.07434
 ridge alpha: 0.01743, val corr: 0.07434
 ridge alpha: 0.03039, val corr: 0.07434
 ridge alpha: 0.05298, val corr: 0.07433
 ridge alpha: 0.09237, val corr: 0.07432
 ridge alpha: 0.16103, val corr: 0.07430
 ridge alpha: 0.28072, val corr: 0.07427
 ridge alpha: 0.48939, val corr: 0.07422
 ridge alpha: 0.85317, val corr: 0.07415
 ridge alpha: 1.48735, val corr: 0.07406
 ridge alpha: 2.59294, val corr: 0.07397
 ridge alpha: 4.52035, val corr: 0.07398
 ridge alpha: 7.88046, val corr: 0.07427
 ridge alpha: 13.73824, val corr: 0.07512
 ridge alpha: 23.95027, val corr: 0.07686
 ridge alpha: 41.75319, val corr: 0.07981
 ridge alpha: 72.78954, val corr: 0.08418
 ridge alpha: 126.89610, val corr: 0.09004
 ridge alpha: 221.22163, val corr: 0.09722
 ridge alpha: 385.66204, val corr: 0.10525
 ridge alpha: 672.33575, val corr: 0.11324
 ridge alpha: 1172.10230, val corr: 0.12005
 ridge alpha: 2043.35972, val corr: 0.12481
 ridge alpha: 3562.24789, val corr: 0.1

In [9]:
# further fine-tuning the ridge regression regularization parameter
ensemble.fit_ridge(X_train, y_train, X_val, y_val, range(5000, 9000, 200))

 ridge alpha: 5000.00000, val corr: 0.12781
 ridge alpha: 5200.00000, val corr: 0.12784
 ridge alpha: 5400.00000, val corr: 0.12785
 ridge alpha: 5600.00000, val corr: 0.12786
 ridge alpha: 5800.00000, val corr: 0.12786
 ridge alpha: 6000.00000, val corr: 0.12786
 ridge alpha: 6200.00000, val corr: 0.12785
 ridge alpha: 6400.00000, val corr: 0.12784
 ridge alpha: 6600.00000, val corr: 0.12783
 ridge alpha: 6800.00000, val corr: 0.12781
 ridge alpha: 7000.00000, val corr: 0.12779
 ridge alpha: 7200.00000, val corr: 0.12777
 ridge alpha: 7400.00000, val corr: 0.12775
 ridge alpha: 7600.00000, val corr: 0.12773
 ridge alpha: 7800.00000, val corr: 0.12770
 ridge alpha: 8000.00000, val corr: 0.12767
 ridge alpha: 8200.00000, val corr: 0.12765
 ridge alpha: 8400.00000, val corr: 0.12762
 ridge alpha: 8600.00000, val corr: 0.12759
 ridge alpha: 8800.00000, val corr: 0.12756
Best ridge alpha: 5800.00000, val corr: 0.12786


The final validation score is 0.12786, which is an improvement over the previous model that used aggregated features from each cluster.

In [10]:
ensemble.refit_full(df_train.drop(columns='label'), df_train['label'])

y_test_pred = ensemble.predict(df_test)
y_test_pred_pd = pd.Series(y_test_pred, name = 'prediction')
y_test_pred_pd.index = df_test.index
y_test_pred_pd.to_csv('../data/output/submission6.csv')

Refitting model 1/60 on full data...
Refitting model 2/60 on full data...
Refitting model 3/60 on full data...
Refitting model 4/60 on full data...
Refitting model 5/60 on full data...
Refitting model 6/60 on full data...
Refitting model 7/60 on full data...
Refitting model 8/60 on full data...
Refitting model 9/60 on full data...
Refitting model 10/60 on full data...
Refitting model 11/60 on full data...
Refitting model 12/60 on full data...
Refitting model 13/60 on full data...
Refitting model 14/60 on full data...
Refitting model 15/60 on full data...
Refitting model 16/60 on full data...
Refitting model 17/60 on full data...
Refitting model 18/60 on full data...
Refitting model 19/60 on full data...
Refitting model 20/60 on full data...
Refitting model 21/60 on full data...
Refitting model 22/60 on full data...
Refitting model 23/60 on full data...
Refitting model 24/60 on full data...
Refitting model 25/60 on full data...
Refitting model 26/60 on full data...
Refitting model 27/60